Adding support for call from memory gadgets ( for ret2csu )

tomgond opened the pull request:
**Overview**:

This PR adds support for ret2csu style gadgets by supporting call gadgets that get the target destination from memory location. I've put special care not to break any existing functionality or code. This works and tested for x64 & MIPS. 
I've done tests to see how much it improves the ability to call a function with 2 parameters by running all the x64 and mips binaries under `angr/binaries/tests`
The results are:

| Arch | Previous | This PR | Improvement |
| ---- | -------- | ------- | ----------- |
| x64  | 140      | 162     | 15%         |
| MIPS | 0        | 36      | N/A         |

**Making sure nothing breaks:**

In order to support this type of gadget I've had to relax some previous restrictions. I made sure to remove the restrictions only for the new gadget types (when possible). I've ran all the test (and added new) and made sure they pass. I've also compared against all the binaries in angr/binaries project. And made sure this PR only adds binaries in which we can call a function with 2 arguments (and does not prevent from any binary that previously could do this call doing it with the PR).

**Main Changes:**

`gadget_analyzer.py` - Add new ctrl_type (`mem_control`) when the ip destination is read from memory.  Add new transit type (`call_from_mem`) for `mem_control` when the `mem_control` instruction is a call.  Added an attribute in gadget object when the transit type is `call_from_mem` called `mem_target_regs` - This specifies for a call, what registers we need to set in order to control the call destination.  Relaxed some restrictions on stack change. Also previous stack changes seemed to be unsigned. Added sign cast to handle negative values better.

`builder.py` - Consider values from `mem_target_regs` for `_build_reg_setting_chain` since we will have to set their value in order to perform a call to our destination

`func_caller.py` - Since we need to call an address read from the program memory, we will have to find a pointer to this address some where in the program. Added code to handle this. In `_func_call` added code to handle chains where the transit_type of the last gadgets are call from mem

`reg_setter.py` - 
 - In `verify` - allow symbolic registers is this is a a call_from_mem gadget.  
 - In `run` - in case we didn't find chain in `_find_reg_setting_gadgets` attempt to run again with `use_partial_controllers` = True
 - In `_find_relevant_gadgets` - Allow call_from_mem transit type gadgets.
 - In `_find_reg_setting_gadgets` - Exported the graph search into another function. If we don't find a chain in the first run, run it again allowing call_from_mem gadgets (permissive_search=True)
 - In `_check_if_sufficient_partial_control` - Changed the check if satisfiable 

`rop_gadget.py` - Added attributes related to call_from_mem gadgets

`rop_utils.py` - `find_mem_access_control` - given memory access formula (e.g `call [rax+8*rbx]` ) determine what registers we need to zero out, and what registers we want to change in order to read from desired memory location. This only works for formula of this type (`a+bx`) and will result in rax = control, rbx = 0.

`tests` - 
- moved `execute_chain` from test_rop to rop_utils, and added control of the goal addr.
- ignored negative (which were unsigned) RopGadgets in `test_rop_i385_cgc`
- added a test for mips and x64 for ret2csu style gadgets. 



Issues:

- The PR is not synced with the latest angr/angrop main since there was a major commit which I could not merge. Hope someone could salvage something from this PR.
- MIPS function calls with ret2csu only work when `needs_return = False`
- I think previous check in `gadget_analyzer.py` for `if gadget.stack_change < 0` was unsigned so always true. after making  `stack_change` value signed some test failed. I've allowed some test with negative (unsigned) stack_change value (`test_rop_i386_cgc`)
- test_arm_func_call sometimes fails with `angrop.errors.RopException: Does not get to an unconstrained successor` but this happens regardless of my code [87](https://github.com/angr/angrop/actions/runs/12898858675/job/35966781773#step:5:88)`

Kyle-Kyle commented:
If I understand the PR correctly. The main things are:
1. primitive support for gadgets ending with call
2. support for gadgets getting PC from a memory region
3. using gadgets ending with calls to set registers (thus ret2csu style)
Is that correct?

Due to the massive conflicts between the current PR and the master branch. I don't think I can merge the current PR. I'll try to rebase it and see whether I can resolve the conflicts. If not, do you mind if I rewrite it? (this way you may not get credited in the commit logs)

If I rewrite it, I'd want to make negative stack-change gadgets and jmp_mem gadgets two features. So ret2csu becomes just a special case of the combination of these two.

tomgond commented:
You are correct about 1 & 2. 
About 3 - I didn't add support for using a `call` to set registers. I've mainly used previous gadgets and existing reg_setter logic to set the registers for the call using 1 & 2. 

I don't mind you rewriting it, it will be great having this support in angrop in any way possible (a mention somewhere would be cool though).

I think re-writing it to make it a special case of negative stack change and jmp_mem is a great way to go. I tried not to change too much code so I won't break anything so I didn't go this way (initially I also added support for jmp_from_mem, but it seemed like it didn't help much enabling func_call with multiple arguments based on angr/binaries dataset, so I removed it).

Kyle-Kyle commented:
Sounds good! I'll look into it and credit you in the code :D

Kyle-Kyle commented:
merged :)
https://github.com/angr/angrop/pull/128

Kyle-Kyle commented:
I added the acknowledgement here:
https://github.com/angr/angrop/pull/128/files#diff-2f3b5bb6a29b43bfe118b6783ac797efd8533bab6361cfe3dd0cb117815ced94R18
And here is the ret2csu testcase:
https://github.com/angr/angrop/pull/128/files#diff-add18372f722092ba56ec70ef9cd474a14e028d17b2ee6b84e01e86fb3971783R536

tomgond commented:
This is amazing. Thanks a lot!
